relations/src/xor.rs

```diff
@@ -21,69 +7,86 @@ use crate::{
 ///  - 1 private witness (b | `private_xoree`)
 ///  - 1 constant        (c | `result`)
 /// such that: a ^ b = c.
-#[derive(Copy, Clone, Eq, PartialEq, Hash, Debug)]
-pub struct XorRelation<S: State> {
-    // ToDo: Especially for Groth16, it is better to provide public input as a field element.
-    // Otherwise, we have to provide it to circuit bit by bit.
-    pub public_xoree: Option<u8>,
-    pub private_xoree: Option<u8>,
+#[snark_relation]
+mod relation {
+    use ark_r1cs_std::{alloc::AllocVar, eq::EqGadget, uint8::UInt8};
 
-    pub result: u8,
+    use crate::byte_to_bits;
 
-    _phantom: PhantomData<S>,
-}
-
-impl XorRelation<NoInput> {
-    pub fn without_input(result: u8) -> Self {
-        Self {
-            public_xoree: None,
-            private_xoree: None,
-            result,
-            _phantom: PhantomData,
-        }
-    }
-}
-
-impl XorRelation<OnlyPublicInput> {
-    pub fn with_public_input(public_xoree: u8, result: u8) -> Self {
-        Self {
-            public_xoree: Some(public_xoree),
-            private_xoree: None,
-            result,
-            _phantom: PhantomData,
-        }
-    }
-}
+    #[circuit_field]
+    pub type CircuitField = crate::CircuitField;
```

Comment:
Do you expect we would ever have a different type for this associated type?

Reply:
:man_shrugging: I really don't know...